wallet/src/state.rs

```diff
@@ -76,6 +80,46 @@ impl ClientState {
             .finalize(rng)
     }
 
+    /// Get a mapping of asset denominations to unspent notes.
+    pub fn notes_by_asset_denomination(&self) -> HashMap<String, Vec<Note>> {
+        let mut notemap = HashMap::<String, Vec<Note>>::new();
+
+        for (_commitment, note) in self.unspent_set.iter() {
+            // Any notes we have in the unspent set we will have the corresponding denominations
+            // for since the notes and asset registry are both part of the sync.
+            let asset_denom = self
+                .asset_registry
+                .get(&note.asset_id())
+                .expect("all asset IDs should have denominations stored locally");
+
+            let new_notes = match notemap.get(asset_denom) {
+                Some(current_notes) => {
+                    let mut new_notes: Vec<Note> =
+                        current_notes.iter().map(|note| note.clone()).collect();
+                    new_notes.push(note.clone());
+                    new_notes.to_vec()
+                }
+                None => {
+                    vec![note.clone()]
+                }
+            };
```

Comment:
non-blocking: this could be cleaner using the `Entry` [API](https://doc.rust-lang.org/stable/std/collections/hash_map/enum.Entry.html#), maybe sth like
```
notemap.entry(asset_denom).or_insert(Vec::new()).push(note.clone());
```